pkg/plugin/manager/manager.go

```diff
@@ -4,161 +4,249 @@
 	if err != nil {
 		return nil, err
 	}
 
-	if !exists {
-		return nil, fmt.Errorf("the config file doesn't exist. Please run cofidectl init")
+	dsName := cfg.Plugins.GetDataSource()
+	if dsName == "" {
+		return nil, errors.New("plugin name cannot be empty")
 	}
 
-	cfg, err := pm.configLoader.Read()
+	// Check if an in-process data source implementation has been requested.
+	if dsName == LocalDSPluginName {
+		ds, err := local.NewLocalDataSource(pm.configLoader)
+		if err != nil {
+			return nil, err
+		}
+		if err := ds.Validate(ctx); err != nil {
+			return nil, err
+		}
+		pm.source = ds
+		return pm.source, nil
+	}
+
+	if err := pm.loadGrpcPlugin(ctx, dsName, cfg.Plugins); err != nil {
+		return nil, err
+	}
+	return pm.source, nil
+}
+
+// GetProvision returns the provision plugin, loading it if necessary.
+func (pm *PluginManager) GetProvision(ctx context.Context) (provision.Provision, error) {
+	if pm.provision != nil {
+		return pm.provision, nil
+	}
+	return pm.loadProvision(ctx)
+}
+
+// loadProvision loads the provision plugin, which may be an in-process or gRPC plugin.
+func (pm *PluginManager) loadProvision(ctx context.Context) (provision.Provision, error) {
+	if pm.provision != nil {
+		return nil, errors.New("provision plugin has already been loaded")
+	}
+
+	cfg, err := pm.readConfig()
 	if err != nil {
 		return nil, err
 	}
 
-	if cfg.DataSource == "" {
-		return nil, errors.New("plugin name cannot be empty")
+	provisionName := cfg.Plugins.GetProvision()
+	if provisionName == "" {
+		return nil, errors.New("provision plugin name cannot be empty")
 	}
 
-	var ds cofidectl_plugin.DataSource
-	var client *go_plugin.Client
-	switch cfg.DataSource {
-	case LocalPluginName:
-		ds, err = local.NewLocalDataSource(pm.configLoader)
-		if err != nil {
-			return nil, err
-		}
-	default:
-		logger := hclog.New(&hclog.LoggerOptions{
-			Name:   "plugin",
-			Output: os.Stdout,
-			Level:  hclog.Error,
-		})
-
-		client, ds, err = pm.loadGrpcPlugin(logger, cfg.DataSource)
-		if err != nil {
+	// Check if an in-process provision implementation has been requested.
+	if provisionName == SpireHelmProvisionPluginName {
+		spireHelm := spirehelm.NewSpireHelm(nil)
+		if err := spireHelm.Validate(ctx); err != nil {
 			return nil, err
 		}
+		pm.provision = spireHelm
+		return pm.provision, nil
 	}
 
-	if err := ds.Validate(); err != nil {
-		if client != nil {
-			client.Kill()
-		}
+	if err := pm.loadGrpcPlugin(ctx, provisionName, cfg.Plugins); err != nil {
 		return nil, err
 	}
-	pm.source = ds
-	pm.client = client
-	return ds, nil
+	return pm.provision, nil
+}
+
+// loadGrpcPlugin loads a gRPC plugin.
+// The gRPC plugin may provide one or more cofidectl plugins (e.g. data source, provision), and
+// all cofidectl plugins configured to use this gRPC plugin will be loaded in a single plugin
+// client and server process.
+func (pm *PluginManager) loadGrpcPlugin(ctx context.Context, pluginName string, pluginCfg *pluginspb.Plugins) error {
```

Comment:
Nit: Could this be renamed to `loadGrpcPlugin` perhaps, following this [suggestion](https://go.dev/wiki/CodeReviewComments#initialisms)?

Reply:
oops, missed this one